src/lib/tls/tls13_pqc/kex_to_kem_adapter.cpp

```diff
@@ -0,0 +1,266 @@
+   #include <botan/dl_group.h>
+#endif
+
+#if defined(BOTAN_HAS_ECDH)
+   #include <botan/ecdh.h>
+#endif
+
+#if defined(BOTAN_HAS_CURVE_25519)
+   #include <botan/curve25519.h>
+#endif
+
+namespace Botan::TLS {
+
+namespace {
+
+/**
+ * Up-cast a key agreement public key to its concrete algorithm type
+ * to perform low-level operations that are not covered by a generic
+ * interface in Public_Key.
+ */
+std::variant<std::reference_wrapper<const DH_PublicKey>,
+             std::reference_wrapper<const ECDH_PublicKey>,
+             std::reference_wrapper<const Curve25519_PublicKey>>
+as_kex_public_key(const Public_Key& kex_public_key) {
+   BOTAN_ASSERT_NOMSG(kex_public_key.supports_operation(PublicKeyOperation::KeyAgreement));
+
+   try {
+#if defined(BOTAN_HAS_ECDH)
+      if(kex_public_key.algo_name() == "ECDH") {
+         return dynamic_cast<const ECDH_PublicKey&>(kex_public_key);
+      }
+#endif
+
+#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
+      if(kex_public_key.algo_name() == "DH") {
+         return dynamic_cast<const DH_PublicKey&>(kex_public_key);
+      }
+#endif
+
+#if defined(BOTAN_HAS_CURVE_25519)
+      if(kex_public_key.algo_name() == "Curve25519") {
+         return dynamic_cast<const Curve25519_PublicKey&>(kex_public_key);
+      }
+#endif
+   } catch(const std::bad_cast&) {
+      throw Invalid_Argument(fmt("public key claimed to be '{}' but down-cast failed", kex_public_key.algo_name()));
+   }
+
+   throw Not_Implemented(fmt("Cannot use unknown key agreement public key of type '{}' in the hybrid KEM key",
+                             kex_public_key.algo_name()));
+}
+
+/**
+ * This helper converts a key agreement public key into its raw public
+ * value. In contrast to the private key (cf. PK_Key_AgreementKey) there
+ * is no generic interface class to do this.
+ */
+std::vector<uint8_t> kex_public_value(const Public_Key& kex_public_key) {
+   return std::visit([](const auto& kex_pub_key) { return kex_pub_key.get().public_value(); },
+                     as_kex_public_key(kex_public_key));
+}
+
+/**
+ * This helper determines the length of the agreed-upon value depending
+ * on the key agreement public key's algorithm type. It would be better
+ * to get this value via PK_Key_Agreement::agreed_value_size(), but
+ * instantiating a PK_Key_Agreement object requires a PrivateKey object
+ * which we don't have (yet) in the context this is used.
+ *
+ * TODO: Find a way to get this information without duplicating those
+ *       implementation details of the key agreement algorithms.
+ */
+size_t kex_shared_key_length(const Public_Key& kex_public_key) {
+   return std::visit(
+      overloaded{[](const ECDH_PublicKey& ecdh_public_key) { return ecdh_public_key.domain().get_p_bytes(); },
+                 [](const DH_PublicKey& dh_public_key) { return dh_public_key.group().p_bytes(); },
+                 [](const Curve25519_PublicKey&) { return size_t(32) /* TODO: magic number */; }},
+      as_kex_public_key(kex_public_key));
+}
```

Comment:
It would be great if we could access certain basic information of specific algorithms statically.

Reply:
Statically might be a bit tricky (given that DH/ECDH are parameterized by the group), but we easily could/should add a function on `PK_Key_Agreement_Key` which specifies what the output length is when you do a raw key agreement.

Reply:
Thought I had left a comment on this earlier? Hm.

Statically might be difficult given how DH/ECDH are defined over arbitrary groups, but it seems pretty possible to have something like `virtual size_t raw_shared_key_length() = 0` the challenge is on what type to put it since `PK_Key_Agreement_Key` is only for the private keys ...

Conceivably we could have something on `Public_Key` like so

````
/**
* Return the length of the relevant artifact of this key.
* 
* Throws Not_Implemented if the key does not support an operation of this type.
*
* For KeyAgreement, returns the raw (non-KDF) output length.
*/
virtual size_t output_length(PublicKeyOperation op) = 0
```